pkg/suggest/rules_yaml.go

```diff
@@ -0,0 +1,247 @@
+// Copyright 2023 The Okteto Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package suggest
+
+import (
+	"errors"
+	"fmt"
+	"github.com/okteto/okteto/pkg/model"
+	"reflect"
+	"regexp"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+	"strings"
+)
+
+// getManifestSuggestionRules returns a collection of rules aiming to improve the error message for the okteto manifest.
+func getManifestSuggestionRules() []ruleInterface {
+	rules := []ruleInterface{
+		addYamlParseErrorHeading(),
+		addUrlToManifestDocs(),
+	}
+
+	// add levenshtein rules to suggest similar field names
+	manifestKeys := getStructKeys(model.Manifest{})
+	manifestKeys["model.manifestRaw"] = manifestKeys["model.Manifest"]
+	manifestKeys["model.buildInfoRaw"] = manifestKeys["model.BuildInfo"]
+	manifestKeys["model.DevRC"] = manifestKeys["model.Dev"]
+
+	for structName, structKeywords := range manifestKeys {
+		for _, keyword := range structKeywords {
+			rule := newLevenshteinRule(fmt.Sprintf(`field (\w+) not found (in type|into) %s`, structName), keyword)
+			rules = append(rules, rule)
+		}
+	}
+
+	rules = append(rules,
+		// invalid properties
+		fieldsNotExistingRule(),
+
+		// struct names
+		newStrReplaceRule("in type model.manifestRaw", "the okteto manifest"),
+		newStrReplaceRule("in type model.ManifestBuild", "the 'build' section"),
+		newStrReplaceRule("into model.ManifestBuild", "into a 'build' object"),
+		newStrReplaceRule("in type model.buildInfoRaw", "the 'build' object"),
+
+		// yaml types
+		newStrReplaceRule(yaml.NodeTagSeq, "list"),
+		newStrReplaceRule(yaml.NodeTagString, "string"),
+		newStrReplaceRule(yaml.NodeTagBool, "boolean"),
+		newStrReplaceRule(yaml.NodeTagInt, "integer"),
+		newStrReplaceRule(yaml.NodeTagFloat, "float"),
+		newStrReplaceRule(yaml.NodeTagMap, "object"),
+
+		// misc
+		newStrReplaceRule("yaml: unmarshal errors:\n", ""),
+		newStrReplaceRule("invalid manifest: ", ""),
+		indentNumLines(),
+	)
+
+	return rules
+}
+
+func isYamlErrorWithoutLinkToDocs(err error) bool {
+	return isYamlError(err) && !hasLinkToDocs(err)
+}
+
+// hasLinkToDocs checks if the error already contains a link to our docs
+func hasLinkToDocs(err error) bool {
+	return strings.Contains(err.Error(), "https://www.okteto.com/docs")
+}
+
+// isYamlError check if the error is related to YAML unmarshalling
+func isYamlError(err error) bool {
+	if err == nil {
+		return false
+	}
```

Comment:
this is to make sure we don't apply the rule if the pattern provides creates an invalid regex, which returns that `err`

Reply:
it's a defensive way to prevent from interrupting the execution or returning any error to the user considered that this is supposed to be transparent for the user